File: src/org/cerion/symcalc/expression/function/integer/IntegerDigits.java
package org.cerion.symcalc.expression.function.integer;

import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.FunctionExpr;
import org.cerion.symcalc.expression.ListExpr;
import org.cerion.symcalc.expression.NumberExpr;
import org.cerion.symcalc.expression.number.IntegerNum;

public class IntegerDigits extends FunctionExpr {

	public IntegerDigits(Expr ...e) {
		super(FunctionType.INTEGER_DIGITS, e);
	}
	
	@Override
	public Expr evaluate() {
		
		if(get(0).isInteger())
		{
			IntegerNum num = (IntegerNum)get(0);
			ListExpr list = new ListExpr();
			
			String s = num.toString();
			for(int i = 0; i < s.length(); i++)
				list.add( NumberExpr.parse("" + s.charAt(i)));
			
			return list;	
		}
 				
		return this;
	}
}

File: src/org/cerion/symcalc/expression/function/integer/IntegerDigits.kt
package org.cerion.symcalc.expression.function.integer

import org.cerion.symcalc.expression.*
import org.cerion.symcalc.expression.function.Function
import org.cerion.symcalc.expression.function.FunctionExpr
import org.cerion.symcalc.expression.number.NumberExpr
import org.cerion.symcalc.expression.number.NumberType

class IntegerDigits(vararg e: Expr) : FunctionExpr(Function.INTEGER_DIGITS, *e) {

    public override fun evaluate(): Expr {
        val num = get(0).asInteger()
        val list = ListExpr()

        val s = num.toString()
        for (i in 0 until s.length)
            list.add(NumberExpr.parse("" + s[i]))

        return list
    }

    override fun validate() {
        validateParameterCount(1)
        validateNumberType(0, NumberType.INTEGER)
    }
}
